src/probe_modules/module_dns.c

```diff
@@ -0,0 +1,1043 @@
+}
+ 
+int dns_global_cleanup(UNUSED struct state_conf *zconf, 
+        UNUSED struct state_send *zsend, UNUSED struct state_recv *zrecv) 
+{
+    if (dns_packet) {
+        free(dns_packet);
+    }
+    dns_packet = NULL;
+
+    if (qname) {
+        free(qname);
+    }
+    qname = NULL;
+
+    return EXIT_SUCCESS;
+}
+
+int dns_init_perthread(void* buf, macaddr_t *src,
+        macaddr_t *gw, UNUSED port_h_t dst_port,
+        UNUSED void **arg_ptr) 
+{
+    memset(buf, 0, MAX_PACKET_SIZE);
+    
+    struct ether_header *eth_header = (struct ether_header *) buf;
+    make_eth_header(eth_header, src, gw);
+    
+    struct ip *ip_header = (struct ip*)(&eth_header[1]);
+    uint16_t len = htons(sizeof(struct ip) + sizeof(struct udphdr) + 
+        dns_packet_len);
+    make_ip_header(ip_header, IPPROTO_UDP, len);
+
+    struct udphdr *udp_header = (struct udphdr*)(&ip_header[1]);
+    len = sizeof(struct udphdr) + dns_packet_len;
+    make_udp_header(udp_header, zconf.target_port, len);
+
+    char* payload = (char*)(&udp_header[1]);
+
+    module_dns.packet_length = sizeof(struct ether_header) + sizeof(struct ip)
+                + sizeof(struct udphdr) + dns_packet_len;
+    assert(module_dns.packet_length <= MAX_PACKET_SIZE);
+
+    memcpy(payload, dns_packet, dns_packet_len);
+
+    return EXIT_SUCCESS;
+}
+
+int dns_make_packet(void *buf, ipaddr_n_t src_ip, ipaddr_n_t dst_ip,
+        uint32_t *validation, int probe_num, UNUSED void *arg) 
+{
+    struct ether_header *eth_header = (struct ether_header *) buf;
+    struct ip *ip_header = (struct ip*) (&eth_header[1]);
+    struct udphdr *udp_header= (struct udphdr *) &ip_header[1];
+
+    ip_header->ip_src.s_addr = src_ip;
+    ip_header->ip_dst.s_addr = dst_ip;
+    udp_header->uh_sport = htons(get_src_port(num_ports, probe_num,
+                     validation));
+    
+    dns_header* dns_header_p = (dns_header*) &udp_header[1];
+    dns_header_p->id = validation[2] & 0xFFFF;
+    
+    ip_header->ip_sum = 0;
+    ip_header->ip_sum = zmap_ip_checksum((unsigned short *) ip_header);
+
+    return EXIT_SUCCESS;
+}
+
+void dns_print_packet(FILE *fp, void* packet) 
+{
+    struct ether_header *ethh = (struct ether_header *) packet;
+    struct ip *iph = (struct ip *) &ethh[1];
+    struct udphdr *udph  = (struct udphdr*) (iph + 4*iph->ip_hl);
+    fprintf(fp, "------------------------------------------------------\n");
+    fprintf(fp, "dns { source: %u | dest: %u | checksum: %#04X }\n",
+        ntohs(udph->uh_sport),
+        ntohs(udph->uh_dport),
+        ntohs(udph->uh_sum));
+    fprintf_ip_header(fp, iph);
+    fprintf_eth_header(fp, ethh);
+    fprintf(fp, "------------------------------------------------------\n");
+}
+
+int dns_validate_packet(const struct ip *ip_hdr, uint32_t len,
+        uint32_t *src_ip, uint32_t *validation)
+{
+    // This does the heavy lifting.
+    if (!udp_validate_packet(ip_hdr, len, src_ip, validation)) {
+        return 0;
+    }
+
```

Comment:
We check (1) below. We save (2) for process packet (as we discussed once upon a time).